### ISSUE TITLE
'test_hinsage_from_generator' test fails

### ISSUE THREAD

**Issue by huonw** [stellargraph/stellargraph#626]
**Describe the bug**

The `test_hinsage_from_generator` test in tests/layer/test_hinsage.py is failing due to

```
>       assert actual == pytest.approx(expected)
E       assert array([[24., 31.],\n       [32., 32.]], dtype=float32) == approx([[26.0 ± 2.6e-05, 29.0 ± 2.9e-05], [32.0 ± 3.2e-05, 31.0 ± 3.1e-05]])
```

In particular, 24 != 26, 31 != 29, and 32 != 31.

**To Reproduce**
1. Go to CI https://buildkite.com/stellar/stellargraph-public/builds/339

```
============================= test session starts ==============================
platform linux -- Python 3.6.10, pytest-5.3.1, py-1.8.1, pluggy-0.13.1
benchmark: 3.2.2 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /workdir, inifile: pytest.ini
plugins: benchmark-3.2.2, cov-2.8.1
collected 334 items                                                            
 
tests/core/test_schema.py .....                                          [  1%]
tests/core/test_stellargraph.py .........................                [  8%]
tests/core/test_types.py ..                                              [  9%]
tests/core/test_utils.py .....                                           [ 11%]
tests/data/test_biased_random_walker.py ............                     [ 14%]
tests/data/test_breadth_first_walker.py .........                        [ 17%]
tests/data/test_converters.py ............                               [ 20%]
tests/data/test_directed_breadth_first_sampler.py ......                 [ 22%]
tests/data/test_edge_splitter.py .........                               [ 25%]
tests/data/test_epgm.py .........                                        [ 28%]
tests/data/test_heterogeneous_breadth_first_walker.py ......             [ 29%]
tests/data/test_metapath_walker.py ......                                [ 31%]
tests/data/test_node_splitter.py ..........                              [ 34%]
tests/data/test_uniform_random_walker.py ......                          [ 36%]
tests/data/test_unsupervised_sampler.py ....                             [ 37%]
tests/layer/test_appnp.py .....                                          [ 39%]
tests/layer/test_attri2vec.py ...                                        [ 40%]
tests/layer/test_cluster_gcn.py .......                                  [ 42%]
tests/layer/test_gcn.py .........                                        [ 44%]
tests/layer/test_graph_attention.py ...................                  [ 50%]
tests/layer/test_graphsage.py ............................               [ 58%]
tests/layer/test_hinsage.py .................F                           [ 64%]
tests/layer/test_link_inference.py .......                               [ 66%]
tests/layer/test_misc.py ...                                             [ 67%]
tests/layer/test_ppnp.py ..                                              [ 67%]
tests/layer/test_rgcn.py ..........                                      [ 70%]
tests/mapper/test_benchmark_generators.py ......                         [ 72%]
tests/mapper/test_cluster_gcn_node_mapper.py ....                        [ 73%]
tests/mapper/test_directed_node_generator.py ..                          [ 74%]
tests/mapper/test_link_mappers.py .......................                [ 81%]
tests/mapper/test_node_mappers.py ....................................   [ 92%]
tests/mapper/test_relational_node_mappers.py .........                   [ 94%]
tests/utils/test_calibration.py ......                                   [ 96%]
tests/utils/test_ensemble.py ........                                    [ 99%]
tests/utils/test_saliency_maps_gat.py .                                  [ 99%]
tests/utils/test_saliency_maps_gcn.py ..                                 [100%]
 
=================================== FAILURES ===================================
_________________________ test_hinsage_from_generator __________________________
 
    def test_hinsage_from_generator():
        G = example_hin_1({"A": 8, "B": 4})
    
        gen = HinSAGENodeGenerator(G, 1, [2, 2], "A", seed=1234)
    
        hs = HinSAGE(
            layer_sizes=[2, 2],
            generator=gen,
            normalize="none",
            kernel_initializer="ones",
            activations=["relu", "relu"],
        )
    
        xin, xout = hs.build()
        model = keras.Model(inputs=xin, outputs=xout)
        actual = model.predict_generator(gen.flow([1, 2]))
        expected = np.array([[26, 29], [32, 31]], dtype=np.float32)
>       assert actual == pytest.approx(expected)
E       assert array([[24., 31.],\n       [32., 32.]], dtype=float32) == approx([[26.0 ± 2.6e-05, 29.0 ± 2.9e-05], [32.0 ± 3.2e-05, 31.0 ± 3.1e-05]])
E        +  where approx([[26.0 ± 2.6e-05, 29.0 ± 2.9e-05], [32.0 ± 3.2e-05, 31.0 ± 3.1e-05]]) = <function approx at 0x7f35c42698c8>(array([[26., 29.],\n       [32., 31.]], dtype=float32))
E        +    where <function approx at 0x7f35c42698c8> = pytest.approx
 
tests/layer/test_hinsage.py:587: AssertionError
```

**Expected behavior**
Should pass.

**Screenshots**
N/A

**Desktop (please complete the following information):**
 - OS: macOS, CI machines (https://buildkite.com/stellar/stellargraph-public/builds/339)

**Data used**
N/A

**Additional context**
N/A

**Comment by huonw** [comment]
Based on comparing the dependencies of https://buildkite.com/stellar/stellargraph-public/builds/334 and https://buildkite.com/stellar/stellargraph-public/builds/339, it looks like this is caused by tensorflow 2.1.0 being released:

```
$ pip install tensorflow==2.1.0
$ pytest tests -k test_hinsage_from_generator
...
tests/layer/test_hinsage.py F
...
>       assert actual == pytest.approx(expected)
E       assert array([[24., 31.],\n       [32., 32.]], dtype=float32) == approx([[26.0 ± 2.6e-05, 29.0 ± 2.9e-05], [32.0 ± 3.2e-05, 31.0 ± 3.1e-05]])
...
===== 1 failed, 333 deselected, 1 warning in 10.82s =====
```

Cf.

```
$ pip install tensorflow==2.0.0
$ pytest tests -k test_hinsage_from_generator
...
tests/layer/test_hinsage.py .
...
==== 1 passed, 333 deselected, 1 warning in 4.42s =====

**Comment by huonw** [comment]
#629 xfailed the test, so CI isn't broken anymore, but the underlying issue still exists.

**Comment by timpitman** [comment]
Looking into tensorflow 2.1 issues, I can't get this test working even with tensorflow==2.0.0

**Comment by kieranricardo** [comment]
@timpitman sure thing :)

**Comment by kieranricardo** [comment]
This is really bizarre, I've reverted to an old version of `stellargraph` (with tf 1.14) and using the original `test_hin_1` function I'm still getting the same result.

**Comment by kieranricardo** [comment]
I haven't tracked down the cause, but it looks like the issue is that the randomly sampled batch is different to what was initially tested (despite have the same seed).

Using the old  `test_hin_1` function the ~usually~ following works:

```
expected = np.array([[26, 29], [32, 31]], dtype=np.float32)
for i in range(1000):
    actual = model.predict_generator(gen.flow([1, 2]))
    if np.isclose(actual, expected).all():
        print("woohoo!")
        break
    elif i == 999:
        print("dangit")
```

Comparing exact values from a random process seems brittle to me, also it seems likely that the test was initially written by running Hinsage with that random seed and using that value in test (which doesn't seem valid)... I propose that we replace this test by comparing the output of hinsage to values calculated with numpy operations on the output of `generator.flow`
